File: src/main/python/AoC2024_10.py
#! /usr/bin/env python3
#
# Advent of Code 2024 Day 10
#

import sys
from collections import deque
from enum import Enum
from enum import auto
from enum import unique

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.grid import Cell
from aoc.grid import IntGrid

Input = IntGrid
Output1 = int
Output2 = int


TEST = """\
89010123
78121874
87430965
96549874
45678903
32019012
01329801
10456732
"""


@unique
class Grading(Enum):
    SCORE = auto()
    RATING = auto()

    def get(self, trails: list[list[Cell]]) -> int:
        match self:
            case Grading.SCORE:
                return len({trail[-1] for trail in trails})
            case Grading.RATING:
                return len(trails)


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return IntGrid.from_strings(list(input_data))

    def solve(self, grid: IntGrid, grading: Grading) -> int:
        def bfs(trail_head: Cell) -> list[list[Cell]]:
            trails = list[list[Cell]]()
            q = deque([[trail_head]])
            while q:
                trail = q.pop()
                nxt = len(trail)
                if nxt == 10:
                    trails.append(trail)
                    continue
                for n in grid.get_capital_neighbours(trail[-1]):
                    if grid.get_value(n) == nxt:
                        q.append([*trail, n])
            return trails

        return sum(
            grading.get(bfs(trail_head))
            for trail_head in grid.get_all_equal_to(0)
        )

    def part_1(self, grid: Input) -> Output1:
        return self.solve(grid, Grading.SCORE)

    def part_2(self, grid: Input) -> Output2:
        return self.solve(grid, Grading.RATING)

    @aoc_samples(
        (
            ("part_1", TEST, 36),
            ("part_2", TEST, 81),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2024, 10)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2024_10.java
import static com.github.pareronia.aoc.Utils.concat;

import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.IntGrid;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

@SuppressWarnings({"PMD.ClassNamingConventions", "PMD.NoPackage"})
public final class AoC2024_10 extends SolutionBase<IntGrid, Integer, Integer> {

    private AoC2024_10(final boolean debug) {
        super(debug);
    }

    public static AoC2024_10 create() {
        return new AoC2024_10(false);
    }

    public static AoC2024_10 createDebug() {
        return new AoC2024_10(true);
    }

    @Override
    protected IntGrid parseInput(final List<String> inputs) {
        return IntGrid.from(inputs);
    }

    private int solve(final IntGrid grid, final Grading grading) {
    	@SuppressWarnings("PMD.AtLeastOneConstructor")
        class BFS {
            private static final int TARGET = 10;

            public List<List<Cell>> bfs(final Cell trailHead) {
                final List<List<Cell>> trails = new ArrayList<>();
                final Deque<List<Cell>> q = new ArrayDeque<>(List.of(List.of(trailHead)));
                while (!q.isEmpty()) {
                    final List<Cell> curr = q.pop();
                    if (curr.size() == TARGET) {
                        trails.add(curr);
                        continue;
                    }
                    final int next = grid.getValue(curr.getLast()) + 1;
                    grid.getCapitalNeighbours(curr.getLast())
                            .filter(n -> grid.getValue(n) == next)
                            .forEach(n -> q.add(concat(curr, n)));
                }
                return trails;
            }
        }

        final BFS bfs = new BFS();
        return grid.getAllEqualTo(0).mapToInt(trailhead -> grading.get(bfs.bfs(trailhead))).sum();
    }

    @Override
    public Integer solvePart1(final IntGrid grid) {
        return solve(grid, Grading.SCORE);
    }

    @Override
    public Integer solvePart2(final IntGrid grid) {
        return solve(grid, Grading.RATING);
    }

    enum Grading {
        SCORE,
        RATING;

        public int get(final List<List<Cell>> trails) {
            return switch (this) {
                case SCORE -> (int) trails.stream().map(List::getLast).distinct().count();
                case RATING -> trails.size();
            };
        }
    }

    @Samples({
        @Sample(method = "part1", input = TEST, expected = "36"),
        @Sample(method = "part2", input = TEST, expected = "81"),
    })
    public static void main(final String[] args) throws Exception {
        create().run();
    }

    private static final String TEST =
            """
            89010123
            78121874
            87430965
            96549874
            45678903
            32019012
            01329801
            10456732
            """;
}
